### DISCUSSION TITLE
How to increment a column without general "update" access?

### DISCUSSION THREAD

**Discussion by lucaslcode** [supabase/supabase#4364] [Questions] [answered]
Hello - wondering if the following use case is possible.

I want to be able to call an api endpoint to increment a count in a particular row.  
I have created a procedure and can call it: `let { data, error } = await supabase.rpc('increment', {...})` (or `POST /rpc/increment`).

Since the incrementing is public access, I don't want anon users to be able to do `PATCH /my_table?id=1`  `{"count": 9999999}`  
However, I don't think row-level security works for procedures (in fact, I'm not sure how it is applied here).

Thanks for any tips!

**Comment by GaryAustin1** [comment, marked as answer]
Unless you use "security definer" in your function it will respect RLS on any tables it accesses.  The function will not be able to increment the row if the user can't read/write (I assume increment function does both) that row per RLS.

If you want the increment to be callable by anyone then you would have RLS on your table to stop normal access to the table and use "security definer" in the function so it can bypass RLS when called.  Security Definer functions take on permission of the owner which would bypass RLS if you create the function in the UI.

**Reply by lucaslcode** [reply]
My issue is, I don't want to set UPDATE access on the rows - the procedure will be run by unauthenticated users, and I don't want them to be able to directly update the count.

I think I can probably use privileges instead.

**Reply by GaryAustin1** [reply]
Yes the second part of my answer covers that.  Use RLS to stop updates or what ever you don't want thru normal calls.  A security definer function can be run by anyone, but only does the increment.

**Reply by lucaslcode** [reply]
Oh I see. Thanks for the help!

**Comment by steve-chavez** [comment]
Note that you can remove the default GRANTs from anon and only allow a security definer function do the update:

```sql
alter default privileges in schema public revoke all on tables from anon;
alter default privileges in schema public revoke all on sequences from anon;

revoke update on all tables in schema public from anon;
-- you could keep this scoped to `my_table` instead of all tables
```

(Related discussion regarding `anon` default privileges: https://github.com/supabase/supabase/discussions/1606)
